tests/end-to-end.py

```diff
@@ -1147,6 +1151,89 @@ def __init__(
+      temp_directory (str): name of a temporary directory.
+
+    Returns:
+      bool: True if he hashes files are identical.
+    """
+    hashes_file_path = os.path.join(temp_directory, test_definition.hashes_file)
+
+    result = False
+    if test_definition.reference_hashes_file:
+      reference_hashes_file_path = test_definition.reference_hashes_file
+      if self._test_references_path:
+        reference_hashes_file_path = os.path.join(
+            self._test_references_path, reference_hashes_file_path)
+
+      if not os.path.exists(reference_hashes_file_path):
+        logging.error('No such reference hashes file: {0:s}'.format(
+            reference_hashes_file_path))
+        return False
+
+      with open(reference_hashes_file_path, 'r') as reference_hashes_file:
+        with open(hashes_file_path, 'r') as hashes_file:
+          # Hack to remove paths in the hashes that are different when running
+          # the tests under UNIX and Windows.
+          reference_hashes_list = []
+          for line in reference_hashes_file.readlines():
+            reference_hashes_list.append(line)
+
+          hashes_list = []
+          for line in hashes_file:
+            hashes_list.append(line)
+
+          differences = list(difflib.unified_diff(
+              reference_hashes_list, hashes_list,
+              fromfile=reference_hashes_file_path, tofile=hashes_file_path))
+
+      if differences:
+        differences_output = []
+        for difference in differences:
+          differences_output.append(difference)
+        differences_output = '\n'.join(differences_output)
+        logging.error('Differences: {0:s}'.format(differences_output))
+
+      if not differences:
+        result = True
+
+    return result
+
+  def _CreateHashesFile(self, test_definition, temp_directory):
+    """Calculates the SHA-256 hashes of the exported files.
+
+    Args:
+      test_definition (TestDefinition): test definition.
+      temp_directory (str): name of a temporary directory.
+    """
+    hashes_file_path = os.path.join(temp_directory, test_definition.hashes_file)
+    exported_files_path = os.path.join(temp_directory, 'export')
+
+    with open(hashes_file_path, 'wt', encoding='utf-8') as hashes_file_object:
+      for path, _, filenames in os.walk(exported_files_path):
+        for filename in filenames:
+          file_path = os.path.join(path, filename)
+          if not os.path.isfile(file_path):
+            continue
+
+          hash_context = hashlib.sha256()
+          with open(file_path, 'rb') as file_object:
+            data = file_object.read(self._READ_BUFFER_SIZE)
+            while data:
+              hash_context.update(data)
+              data = file_object.read(self._READ_BUFFER_SIZE)
+
+          relative_file_path = file_path[len(temp_directory):]
+          hashes_file_object.write('{0:s}\t{1:s}\n'.format(
+              hash_context.hexdigest(), relative_file_path))
+
+    if os.path.exists(hashes_file_path):
```

Comment:
should we raise if hashes_file_path doesn't exist, or do we expect _CreateHashesFile to be called with no an empty {temp_directory}/export?

Reply:
IMHO no need to raise here, if the path is missing this will fail the compare step.